components/server/ee/src/prebuilds/prebuild-manager.ts

```diff
@@ -120,18 +124,43 @@ export class PrebuildManager {
         const span = TraceContext.startSpan("retriggerPrebuild", ctx);
         span.setTag("workspaceId", workspaceId);
         try {
+            const workspacePromise = this.workspaceDB.trace({ span }).findById(workspaceId);
+            const prebuildPromise = this.workspaceDB.trace({ span }).findPrebuildByWorkspaceID(workspaceId);
             const runningInstance = await this.workspaceDB.trace({ span }).findRunningInstance(workspaceId);
             if (runningInstance !== undefined) {
                 throw new WorkspaceRunningError('Workspace is still runnning', runningInstance);
             }
             span.setTag("starting", true);
-            const workspace = await this.workspaceDB.trace({ span }).findById(workspaceId);
+            const workspace = await workspacePromise;
             if (!workspace) {
                 console.error('Unknown workspace id.', { workspaceId });
                 throw new Error('Unknown workspace ' + workspaceId);
             }
+            const prebuild = (await prebuildPromise)!;
             await this.workspaceStarter.startWorkspace({ span }, workspace, user);
-            return { wsid: workspace.id, done: false };
+            return { prebuildId: prebuild.id, wsid: workspace.id, done: false };
+        } catch (err) {
+            TraceContext.logError({ span }, err);
+            throw err;
+        } finally {
+            span.finish();
+        }
+    }
+
+    async cancelPrebuild(ctx: TraceContext, user: User, prebuild: PrebuiltWorkspace): Promise<void> {
+        const span = TraceContext.startSpan("cancelPrebuild", ctx);
+        span.setTag("prebuildId", prebuild.id);
+        try {
+            if (prebuild.state === 'aborted' || prebuild.state === 'timeout' || prebuild.state === 'available') {
+                return;
+            }
+            prebuild.state = 'aborted';
+            prebuild.error = 'Cancelled manually';
+            await this.workspaceDB.trace({ span }).storePrebuiltWorkspace(prebuild);
+            await this.messageBus.notifyHeadlessUpdate({ span }, user.id, prebuild.buildWorkspaceId, <HeadlessWorkspaceEvent>{
```

Comment:
how does this play together with state transitions driven by the bridge?
the bridge is receiving status updates from ws manager and does update the prebuild workspace "in parallel". there could be races of `storePrebuiltWorkspace` calls, potentially leading to override the `Aborted` message?

shouldn't the cancellation be a request to the wsman, for which in turn the bridge would to the notifications and updates from this change?

Reply:
Good question, thanks! To be honest, I simply copied the prebuild cancellation logic from `bridge.ts`:

https://github.com/gitpod-io/gitpod/blob/6bc80eb92479e2c9d400d059f5ce2dc56bde304f/components/ws-manager-bridge/ee/src/bridge.ts#L77-L96

and also from how it was implemented in the (now-deleted) `PrebuildQueueMaintainer`:

https://github.com/gitpod-io/gitpod/blob/2aa2db6524c5799ee4546879fe9e33efd8330a4e/components/server/ee/src/prebuilds/prebuild-queue-maintainer.ts#L144-L175

I don't quite understand the full state machine yet, but it sounds like I should look more deeply into it. 👀

Reply:
@jankeromnes, first of https://github.com/gitpod-io/gitpod/pull/5835/files

then, you see it's state propagation from the `ws-manager-bridge` which runs outside of server. which is the whole point: ws-manager running in region A and propagating event from there, but then you want to cancel a prebuild from server in region B. in such cases I'm puzzled about the stream of states a frontend would receive, thus the message for the user would be confusing.

in general, if we want something to happen with the workspace, we're sending requests to the ws-manager of the according region, which is already taken care of by the client connection as far as know. btw, this is what happens for the stop part of this new composite action called cancelPrebuild. I think what needs to be done here is to introduce a CancelPrebuild request in ws-manager, which basically does the stopping but also signals the action to the bridge to do the rest, i.e. update db and push notifications.

Reply:
After our sync discussion, it seems we'll want to remove the new `prebuildManager.cancelPrebuild` call, and instead make `stopWorkspace` also cancel the prebuild (inside the correct `ws-manager`, e.g. with a new flag on the `StopWorkspaceRequest` if necessary).